kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/KopEventManager.java

```diff
@@ -337,23 +346,33 @@ public void process(BiConsumer<String, Long> registerEventLatency,
                         log.info("Tenant {} GroupMetadata delete topics {}, no matching topics {}",
                                 tenant, curDeletedTopics,
                                 Sets.difference(topicsFullNameDeletionsSets, curDeletedTopics));
-
-                        deletedTopics.addAll(curDeletedTopics);
-                    }
-                    if (pendingCoordinators.decrementAndGet() == 0) {
-                        deletedTopics.forEach(deletedTopic -> {
-                            metadataStore.delete(
-                                    getDeleteTopicsPath() + "/" + deletedTopic, Optional.of((long) -1));
-                        });
                     }
                 });
-            } catch (ExecutionException | InterruptedException e) {
-                log.error("DeleteTopicsEvent process have an error", e);
             } finally {
                 registerEventLatency.accept(name(), startProcessTime);
             }
         }
 
+        private void sendDeleteTopics(List<String> topicsDeletions) {
+            Collection<? extends Node> interBrokers = adminManager.getBrokers(kafkaConfig.getInterBrokerListenerName());
+            List<AdminClient> adminClients = Lists.newArrayList();
+            try {
+                interBrokers.forEach(broker -> {
+                    if (!advertisedEndPoint.getHostname().equals(broker.host())
+                            || advertisedEndPoint.getPort() != broker.port()) {
+                        AdminClient adminClient =
+                                AdminClientUtils.getKafkaAdminClient(kafkaConfig, broker.host(), broker.port());
+                        adminClients.add(adminClient);
+                        adminClient.deleteTopics(topicsDeletions);
+                        log.info("Send deleteTopics {} to kop broker {}:{}",
+                                topicsDeletions, broker.host(), broker.port());
+                    }
+                });
```

Comment:
What if there are many brokers? Why does each broker need to process a DELETE_TOPICS request?

Reply:
In kafka, DeleteTopicsRequest will only be sent to the controller, while in kop, the node processing the metadataRequest will return itself to become the controller node.

Reply:
So I want to use this feature of the client and the processing logic of kop, and the kop node that receives the client's delete topic request will forward this request to other kop nodes.

Reply:
Therefore, creating an adminClient for each kop broker can ensure that the corresponding kop broker becomes the controller of this request.

Reply:
The request to update metadata in Kafka is also distributed by the Controller node to all Kafka brokers.

Reply:
Kafka controller is used to manage metadata of the whole cluster, while each Pulsar broker exposes a HTTP endpoint for metadata management based on ZK (metadata store). Pulsar doesn't need to maintain leader and ISRs so the metadata is not as much as Kafka has.

Considering a case that you found sending the request to all brokers brings a performance overhead, maybe you would like to add a controller role in KoP?

Before going so far, we should avoid broker-to-broker communication from the beginning. If we want to share some metadata among the brokers, use the metadata store or create a system topic.

Reply:
I don't want to add the controller role to kop because I also noticed that pulsar also uses PulsarAdmin to forward messages. However, you are right. It is best to use metadata store or create system topic. As you can see, the previous method is based on metadata store.